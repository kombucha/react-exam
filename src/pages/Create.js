import React, { PureComponent } from "react";
import PropTypes from "prop-types";

class Create extends PureComponent {
  state = { name: "" };

  _handleNameChange = ev => this.setState({ name: ev.target.value });

  _handleSubmit = ev => {
    const { history } = this.props;
    ev.preventDefault();
    this.props
      .createJedi(this.state)
      .then(() => {
        history.push("/");
      })
      .catch(err => {
        alert(`Failed to create jedi: ${err}`);
      });
  };

  render() {
    const { name } = this.state;
    return (
      <div className="Create">
        <form onSubmit={this._handleSubmit}>
          <input type="text" value={name} onChange={this._handleNameChange} />
          <button type="submit">Create</button>
        </form>
      </div>
    );
  }
}

Create.propTypes = {
  createJedi: PropTypes.func.isRequired,
};

export default Create;
