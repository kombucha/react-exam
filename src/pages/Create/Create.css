.Create {
  display: flex;
  flex-direction: column;
  align-items: center;
}

.Create__form {
  display: flex;
  flex-direction: column;
}

.Create__terms {
  text-align: right;
}
