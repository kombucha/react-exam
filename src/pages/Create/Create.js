import React, { PureComponent } from "react";
import PropTypes from "prop-types";

import TextInput from "../../components/TextInput";
import Button from "../../components/Button";
import "./Create.css";

class Create extends PureComponent {
  state = { name: "", termsOfServicesOk: false };

  // NB: Could have used a helper library like https://github.com/christianalfoni/formsy-react
  _handleNameChange = ev => this.setState({ name: ev.target.value });
  _handleTermsOfServicesChange = ev => this.setState({ termsOfServicesOk: ev.target.checked });

  _handleSubmit = ev => {
    const { history } = this.props;
    ev.preventDefault();
    this.props
      .createJedi(this.state)
      .then(() => {
        history.push("/");
      })
      .catch(err => {
        alert(`Failed to create jedi: ${err}`);
      });
  };

  render() {
    const { name, termsOfServicesOk } = this.state;
    return (
      <div className="Create">
        <form className="Create__form" onSubmit={this._handleSubmit}>
          <TextInput
            name="name"
            placeholder="Name"
            type="text"
            value={name}
            required
            minLength={3}
            autoFocus
            onChange={this._handleNameChange}
          />
          <p>
            There is no emotion, there is peace. <br />
            There is no ignorance, there is knowledge.<br />
            There is no passion, there is serenity.<br />
            There is no chaos, there is harmony.<br />
            There is no death, there is the Force<br />
          </p>
          <label className="Create__terms" htmlFor="termsOfServices">
            i accept the jedi code
            <input
              type="checkbox"
              id="termsOfServices"
              name="termsOfServices"
              checked={termsOfServicesOk}
              onChange={this._handleTermsOfServicesChange}
            />
          </label>

          <div>
            <Button type="submit" disabled={!termsOfServicesOk}>
              Create
            </Button>
          </div>
        </form>
      </div>
    );
  }
}

Create.propTypes = {
  createJedi: PropTypes.func.isRequired,
  history: PropTypes.object.isRequired,
};

export default Create;
