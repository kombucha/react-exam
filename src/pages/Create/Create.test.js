import React from "react";
import Create from "./Create";
import { shallow } from "enzyme";

it("should render properly", () => {
  const createJedi = jest.fn();
  const history = {};
  const wrapper = shallow(<Create createJedi={createJedi} history={history} />);
  expect(wrapper).toMatchSnapshot();
});

it("should redirect to home after creation", () => {
  let jediCreationPromise;
  const createJedi = jest.fn(jedi => {
    jediCreationPromise = Promise.resolve(jedi);
    return jediCreationPromise;
  });
  const history = { push: jest.fn() };
  // See https://github.com/airbnb/enzyme/issues/308
  const fakeSubmitEvent = { preventDefault() {} };

  const wrapper = shallow(<Create createJedi={createJedi} history={history} />);
  wrapper.find("form").simulate("submit", fakeSubmitEvent);

  return jediCreationPromise.then(() => {
    expect(createJedi).toHaveBeenCalled();
    expect(history.push).toHaveBeenCalled();
  });
});
