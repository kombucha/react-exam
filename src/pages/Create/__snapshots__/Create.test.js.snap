// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render properly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Create
    createJedi={[Function]}
    history={Object {}}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": null,
    "nodeType": "host",
    "props": Object {
      "children": <form
        className="Create__form"
        onSubmit={[Function]}
>
        <TextInput
                autoFocus={true}
                minLength={3}
                name="name"
                onChange={[Function]}
                placeholder="Name"
                required={true}
                type="text"
                value=""
        />
        <p>
                There is no emotion, there is peace. 
                <br />
                There is no ignorance, there is knowledge.
                <br />
                There is no passion, there is serenity.
                <br />
                There is no chaos, there is harmony.
                <br />
                There is no death, there is the Force
                <br />
        </p>
        <label
                className="Create__terms"
                htmlFor="termsOfServices"
        >
                i accept the jedi code
                <input
                        checked={false}
                        id="termsOfServices"
                        name="termsOfServices"
                        onChange={[Function]}
                        type="checkbox"
                />
        </label>
        <div>
                <Button
                        disabled={true}
                        type="submit"
                >
                        Create
                </Button>
        </div>
</form>,
      "className": "Create",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": null,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <TextInput
            autoFocus={true}
            minLength={3}
            name="name"
            onChange={[Function]}
            placeholder="Name"
            required={true}
            type="text"
            value=""
/>,
          <p>
            There is no emotion, there is peace. 
            <br />
            There is no ignorance, there is knowledge.
            <br />
            There is no passion, there is serenity.
            <br />
            There is no chaos, there is harmony.
            <br />
            There is no death, there is the Force
            <br />
</p>,
          <label
            className="Create__terms"
            htmlFor="termsOfServices"
>
            i accept the jedi code
            <input
                        checked={false}
                        id="termsOfServices"
                        name="termsOfServices"
                        onChange={[Function]}
                        type="checkbox"
            />
</label>,
          <div>
            <Button
                        disabled={true}
                        type="submit"
            >
                        Create
            </Button>
</div>,
        ],
        "className": "Create__form",
        "onSubmit": [Function],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": null,
          "nodeType": "function",
          "props": Object {
            "autoFocus": true,
            "minLength": 3,
            "name": "name",
            "onChange": [Function],
            "placeholder": "Name",
            "required": true,
            "type": "text",
            "value": "",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": null,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              "There is no emotion, there is peace. ",
              <br />,
              "There is no ignorance, there is knowledge.",
              <br />,
              "There is no passion, there is serenity.",
              <br />,
              "There is no chaos, there is harmony.",
              <br />,
              "There is no death, there is the Force",
              <br />,
            ],
          },
          "ref": null,
          "rendered": Array [
            "There is no emotion, there is peace. ",
            Object {
              "instance": null,
              "key": null,
              "nodeType": "host",
              "props": Object {},
              "ref": null,
              "rendered": null,
              "type": "br",
            },
            "There is no ignorance, there is knowledge.",
            Object {
              "instance": null,
              "key": null,
              "nodeType": "host",
              "props": Object {},
              "ref": null,
              "rendered": null,
              "type": "br",
            },
            "There is no passion, there is serenity.",
            Object {
              "instance": null,
              "key": null,
              "nodeType": "host",
              "props": Object {},
              "ref": null,
              "rendered": null,
              "type": "br",
            },
            "There is no chaos, there is harmony.",
            Object {
              "instance": null,
              "key": null,
              "nodeType": "host",
              "props": Object {},
              "ref": null,
              "rendered": null,
              "type": "br",
            },
            "There is no death, there is the Force",
            Object {
              "instance": null,
              "key": null,
              "nodeType": "host",
              "props": Object {},
              "ref": null,
              "rendered": null,
              "type": "br",
            },
          ],
          "type": "p",
        },
        Object {
          "instance": null,
          "key": null,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              "i accept the jedi code",
              <input
                checked={false}
                id="termsOfServices"
                name="termsOfServices"
                onChange={[Function]}
                type="checkbox"
/>,
            ],
            "className": "Create__terms",
            "htmlFor": "termsOfServices",
          },
          "ref": null,
          "rendered": Array [
            "i accept the jedi code",
            Object {
              "instance": null,
              "key": null,
              "nodeType": "host",
              "props": Object {
                "checked": false,
                "id": "termsOfServices",
                "name": "termsOfServices",
                "onChange": [Function],
                "type": "checkbox",
              },
              "ref": null,
              "rendered": null,
              "type": "input",
            },
          ],
          "type": "label",
        },
        Object {
          "instance": null,
          "key": null,
          "nodeType": "host",
          "props": Object {
            "children": <Button
              disabled={true}
              type="submit"
>
              Create
</Button>,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": null,
            "nodeType": "function",
            "props": Object {
              "children": "Create",
              "disabled": true,
              "type": "submit",
            },
            "ref": null,
            "rendered": "Create",
            "type": [Function],
          },
          "type": "div",
        },
      ],
      "type": "form",
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": null,
      "nodeType": "host",
      "props": Object {
        "children": <form
          className="Create__form"
          onSubmit={[Function]}
>
          <TextInput
                    autoFocus={true}
                    minLength={3}
                    name="name"
                    onChange={[Function]}
                    placeholder="Name"
                    required={true}
                    type="text"
                    value=""
          />
          <p>
                    There is no emotion, there is peace. 
                    <br />
                    There is no ignorance, there is knowledge.
                    <br />
                    There is no passion, there is serenity.
                    <br />
                    There is no chaos, there is harmony.
                    <br />
                    There is no death, there is the Force
                    <br />
          </p>
          <label
                    className="Create__terms"
                    htmlFor="termsOfServices"
          >
                    i accept the jedi code
                    <input
                              checked={false}
                              id="termsOfServices"
                              name="termsOfServices"
                              onChange={[Function]}
                              type="checkbox"
                    />
          </label>
          <div>
                    <Button
                              disabled={true}
                              type="submit"
                    >
                              Create
                    </Button>
          </div>
</form>,
        "className": "Create",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": null,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <TextInput
              autoFocus={true}
              minLength={3}
              name="name"
              onChange={[Function]}
              placeholder="Name"
              required={true}
              type="text"
              value=""
/>,
            <p>
              There is no emotion, there is peace. 
              <br />
              There is no ignorance, there is knowledge.
              <br />
              There is no passion, there is serenity.
              <br />
              There is no chaos, there is harmony.
              <br />
              There is no death, there is the Force
              <br />
</p>,
            <label
              className="Create__terms"
              htmlFor="termsOfServices"
>
              i accept the jedi code
              <input
                            checked={false}
                            id="termsOfServices"
                            name="termsOfServices"
                            onChange={[Function]}
                            type="checkbox"
              />
</label>,
            <div>
              <Button
                            disabled={true}
                            type="submit"
              >
                            Create
              </Button>
</div>,
          ],
          "className": "Create__form",
          "onSubmit": [Function],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": null,
            "nodeType": "function",
            "props": Object {
              "autoFocus": true,
              "minLength": 3,
              "name": "name",
              "onChange": [Function],
              "placeholder": "Name",
              "required": true,
              "type": "text",
              "value": "",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": null,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                "There is no emotion, there is peace. ",
                <br />,
                "There is no ignorance, there is knowledge.",
                <br />,
                "There is no passion, there is serenity.",
                <br />,
                "There is no chaos, there is harmony.",
                <br />,
                "There is no death, there is the Force",
                <br />,
              ],
            },
            "ref": null,
            "rendered": Array [
              "There is no emotion, there is peace. ",
              Object {
                "instance": null,
                "key": null,
                "nodeType": "host",
                "props": Object {},
                "ref": null,
                "rendered": null,
                "type": "br",
              },
              "There is no ignorance, there is knowledge.",
              Object {
                "instance": null,
                "key": null,
                "nodeType": "host",
                "props": Object {},
                "ref": null,
                "rendered": null,
                "type": "br",
              },
              "There is no passion, there is serenity.",
              Object {
                "instance": null,
                "key": null,
                "nodeType": "host",
                "props": Object {},
                "ref": null,
                "rendered": null,
                "type": "br",
              },
              "There is no chaos, there is harmony.",
              Object {
                "instance": null,
                "key": null,
                "nodeType": "host",
                "props": Object {},
                "ref": null,
                "rendered": null,
                "type": "br",
              },
              "There is no death, there is the Force",
              Object {
                "instance": null,
                "key": null,
                "nodeType": "host",
                "props": Object {},
                "ref": null,
                "rendered": null,
                "type": "br",
              },
            ],
            "type": "p",
          },
          Object {
            "instance": null,
            "key": null,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                "i accept the jedi code",
                <input
                  checked={false}
                  id="termsOfServices"
                  name="termsOfServices"
                  onChange={[Function]}
                  type="checkbox"
/>,
              ],
              "className": "Create__terms",
              "htmlFor": "termsOfServices",
            },
            "ref": null,
            "rendered": Array [
              "i accept the jedi code",
              Object {
                "instance": null,
                "key": null,
                "nodeType": "host",
                "props": Object {
                  "checked": false,
                  "id": "termsOfServices",
                  "name": "termsOfServices",
                  "onChange": [Function],
                  "type": "checkbox",
                },
                "ref": null,
                "rendered": null,
                "type": "input",
              },
            ],
            "type": "label",
          },
          Object {
            "instance": null,
            "key": null,
            "nodeType": "host",
            "props": Object {
              "children": <Button
                disabled={true}
                type="submit"
>
                Create
</Button>,
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": null,
              "nodeType": "function",
              "props": Object {
                "children": "Create",
                "disabled": true,
                "type": "submit",
              },
              "ref": null,
              "rendered": "Create",
              "type": [Function],
            },
            "type": "div",
          },
        ],
        "type": "form",
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
